// restle/dist/lib/router/utils/check-post.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports['default'] = checkPost;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _restleError = require('restle-error');

var _utilsParseModelName = require('../../utils/parse-model-name');

var _utilsParseModelName2 = _interopRequireDefault(_utilsParseModelName);

/**
 * Checks if the incoming request data is not conflicting.
 *
 * @private
 * @param {Object} req
 * @return {Boolean}
 */

function checkPost(req, res) {
  if (!(req.body && req.body.data && req.body.data.type)) {
    var badRequestError = new _restleError.BadRequestError({
      detail: 'A primary data object with a matching type key is required when creating a resource.'
    });

    return {
      status: badRequestError.status,
      json: badRequestError.serialize()
    };
  }

  if ((0, _utilsParseModelName2['default'])(req.body.data.type) !== (0, _utilsParseModelName2['default'])(req.params.type)) {
    var conflictError = new _restleError.ConflictError({
      first: (0, _utilsParseModelName2['default'])(req.body.data.type),
      second: (0, _utilsParseModelName2['default'])(req.params.type)
    });

    return {
      status: conflictError.status,
      json: conflictError.serialize()
    };
  }
}

module.exports = exports['default'];
